src/mlpack/core/optimizers/adam/adam_impl.hpp

```diff
@@ -113,19 +119,44 @@ double Adam<DecomposableFunctionType>::Optimize(arma::mat& iterate)
     m *= beta1;
     m += (1 - beta1) * gradient;
 
-    v *= beta2;
-    v += (1 - beta2) * (gradient % gradient);
+    if (adaMax)
+    {
+      u *= beta2;
+      arma::mat absGradient = arma::abs(gradient);
+      // Update the exponentially weighted infinity norm.
+      for (size_t row = 0; row != iterate.n_rows; ++row)
+      {
+        for (size_t col = 0; col != iterate.n_cols; ++col)
+        {
+          if (u.at(row, col) < absGradient.at(row, col))
+            u.at(row, col) = absGradient.at(row, col);
+        }
+      }
+    }
+    else
+    {
+      v *= beta2;
+      v += (1 - beta2) * (gradient % gradient);
+    }
 
     const double biasCorrection1 = 1.0 - std::pow(beta1, (double) i);
     const double biasCorrection2 = 1.0 - std::pow(beta2, (double) i);
 
-    /**
-     * It should be noted that the term, m / (arma::sqrt(v) + eps), in the
-     * following expression is an approximation of the following actual term;
-     * m / (arma::sqrt(v) + (arma::sqrt(biasCorrection2) * eps).
-     */
-    iterate -= (stepSize * std::sqrt(biasCorrection2) / biasCorrection1) *
-                m / (arma::sqrt(v) + eps);
+    if (adaMax)
+    {
+      if (biasCorrection1 != 0.0)
+        iterate -= ((stepSize / biasCorrection1) * (m / u));
```

Comment:
I looked over the code, and it looks good; probably using m / (u + eps) might help?